### PR TITLE
Fix some problems with pathlib PosixPath/WindowsPath behavior

### PR DESCRIPTION
**Pull request by camillobruni**
#### Describe the changes
Adding tests to document subtle pyfakefs differences described in issue 1053

#### Tasks
- [x] Unit tests added that reproduce the issue or prove feature is working
- [x] Fix or feature added
- [x] Entry to release notes added
- [x] Pre-commit CI shows no errors
- [x] Unit tests passing
- [ ] For documentation changes: The Read the Docs preview builds and looks as expected

### CODE REVIEW

**Comment by mrbean-bremen** [comment]
@camillobruni - I have added a commit that shall fix most of the issues, but I still got one failing test (test_real_file_with_home) under Posix and Python < 3.12, which I have disabled at the moment. 
Can you please check how this version works for you?

File: pyfakefs/fake_filesystem.py
Comment (kurtmckee) [inline pyfakefs/fake_filesystem.py:283]: Oh boy, I _strongly_ recommend against combining `or` and `and` operators without parentheses here. This is unlikely to work the way people read it.

[Python's operator precedence rules](https://docs.python.org/3/reference/expressions.html#operator-precedence) make it clear that the `and` will be evaluated first, followed by the `or` making this code equivalent to:

```python
self.fs_type == FSType.WINDOWS
or (self.fs_type == FSType.DEFAULT and self._is_windows_fs)
```

I recommend adding grouping parentheses to confirm to readers that this is written -- and will evaluate -- as intended, even if you've written this perfectly and know that it will evaluate as intended.

**Review by kurtmckee — CHANGES_REQUESTED** [review]
I haven't read all of this but wanted to offer feedback for what I did see.

**Comment by mrbean-bremen** [comment]
> I haven't read all of this but wanted to offer feedback for what I did see.

Another pair of eyes is always helpful, especially as I rarely get reviews here. Given enough eyeballs...

**Comment by camillobruni** [comment]
Late reply back from vacation: thanks a lot for these fixes!

File: pyfakefs/fake_pathlib.py
Comment (kurtmckee) [inline pyfakefs/fake_pathlib.py:522]: This catches my eye -- elsewhere (like in `helpers.py`) this is `None`, not an empty string. Is the empty string correct here?

File: pyfakefs/fake_filesystem.py
Comment (mrbean-bremen) [reply pyfakefs/fake_filesystem.py:289]: Good point, I missed that!

File: pyfakefs/fake_filesystem.py
Comment (mrbean-bremen) [reply pyfakefs/fake_filesystem.py:1205]: The problem is not `mypy` - it understands this fine - but `pytype`. I can check if this would help with pytype, anyway, thanks!

File: pyfakefs/fake_pathlib.py
Comment (mrbean-bremen) [reply pyfakefs/fake_pathlib.py:522]: I was just lazy - this way I didn't need to add typing information, and these are set during intializing anyway. But you are right, better to do this consistently.

**Review by kurtmckee — APPROVED** [review]

File: pyfakefs/fake_filesystem.py
Comment (mrbean-bremen) [reply pyfakefs/fake_filesystem.py]: Yes, that heplped with pytype - thanks!

**Comment by mrbean-bremen** [comment]
@camillobruni - I added a workaround to get the failing test running, and I would merge this if you give your ok. Did you have a chance to test it?
Oh, and I forgot the release notes, will add these before merging, of course.

**Comment by camillobruni** [comment]
Cool, double checked locally and I got the tests working with this 👍

**Pull request merged**
